### PR TITLE
Tarmac-Styled Code Generation Support

### PR DESCRIPTION
**Pull request by MerryweatherLost**
Hello!

This PR is meant to add in the feature requested in **https://github.com/jackTabsCode/asphalt/issues/23** related to tarmac code style. I've implemented a solution for this. Utilizing your tests, the emits are as followed so far:

### Luau Emit (4 spaces)
```lua
return {
    bar = {
        ["baz.png"] = "rbxassetid://2",
    },
    ["foo.png"] = "rbxassetid://1",
}
```

### Typescript Emit (4 spaces)
```ts
declare const assets: {
    bar: {
        "baz.png": "rbxassetid://2",
    },
    "foo.png": "rbxassetid://1",
}
export = assets
```

This also adds an extra configuration that is `tarmac`, which will be false by default.

Please feel free to request any changes or provide any edits or questions, and thank you for providing this useful tool!

### CODE REVIEW

**Comment by jackTabsCode** [comment]
Hey, thanks for your work! I haven't had time to look through the full implementation yet, but I am a bit concerned about the organization of code here before that.

Ideally, `codegen.rs` would contain all codegen related code. So `main.rs` should not really worry about what format the code is in. Ideally the main function would send off the `lockfile` to `codegen`, `codegen` would path it based on the desired code generation style, and receive back the string or the bytes needed to write the file, which the main function would do.

Another minor thing-I am not a huge fan of naming the configuration (or the style itself) "tarmac" mode. I would much prefer if it was named "style" and perhaps the options could be "flat" or "nested". We could then say in the readme, if you like the tarmac style, go with nested.

I'm also going to end up putting codegen options, since we have three at this point, in it's own section in the configuration. But this probably belongs in a different PR.

File: src/ast.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        impl fmt::Display for $target {
+            fn fmt(&self, output: &mut fmt::Formatter) -> fmt::Result {
+                let mut stream = ASTStream::new(output, &self.1);
+                ASTFormat::fmt_ast(self, &mut stream)
+            }
+        }
+    };
+}
+
+trait ASTFormat {
```
Comment (jackTabsCode) [inline src/ast.rs:14]: Should probably be `AstFormat`

File: src/ast.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        if is_valid_identifier(self) {
+            write!(output, "{}", self)
+        } else {
+            match output.target {
+                ASTTarget::Lua => write!(output, "[\"{}\"]", self),
+                ASTTarget::Typescript { .. } => write!(output, "\"{}\"", self),
+            }
+        }
+    }
+}
```
Comment (jackTabsCode) [inline src/ast.rs:187]: `AST` should be `Ast`

**Comment by MerryweatherLost** [comment]
Thanks for your feedback! I got a better understanding of what you would intend for your project, and I will be making the requested changes to the best of my ability.

On the `tarmac` naming, `nested` is perfectly fine, and providing a `style` option in the config is more extendable, I agree with you.
I will make those changes momentarily.

**Comment by MerryweatherLost** [comment]
To add to the discussion, would you prefer the config to have a style enum akin to this to replace `tarmac: bool`?

~ `config.rs`
```rs
#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum StyleType {
    Flat,
    Nested
}
```
In the config I would add a `style` option, and it would end up looking like this in `asphalt.toml`

```toml
style = "nested"
```

~ `state.rs`
```rs
pub struct State {
    ...
    pub style: StyleType,
    ...
}
```

**Comment by jackTabsCode** [comment]
Yeah, that should work. Let's make sure it's optional and `Flat` is the default.

File: src/ast.rs (hunk 2 above)
Comment (MerryweatherLost) [reply src/ast.rs]: Should be resolved in: https://github.com/jackTabsCode/asphalt/pull/33/commits/65076f6a46b684dfd3b5b1a965e005c626c4e199

**Comment by MerryweatherLost** [comment]
I've attempted to implement the concept in the previous comment with this commit: https://github.com/jackTabsCode/asphalt/pull/33/commits/17573040e864128f5d296800dc0ea20c5354dce9

**Comment by jackTabsCode** [comment]
Nice. Can we convert `.expect` calls to `.context` to better fit in with the rest of the codebase?

**Comment by jackTabsCode** [comment]
Thank you. I'm not going to release quite yet-going to restructure the config file a little bit, then I'll push out 0.5.0

**Pull request merged**